sdks/go/pkg/beam/io/mongodbio/read.go

```diff
@@ -442,53 +266,53 @@ func (fn *readFn) ProcessElement(
 	}()
 
 	for cursor.Next(ctx) {
-		value, err := decodeDocument(cursor, fn.Type.T)
+		id, value, err := decodeDocument(cursor, fn.Type.T)
 		if err != nil {
 			return err
 		}
 
-		emit(value)
-	}
-
-	return cursor.Err()
-}
+		result := cursorResult{nextID: id}
+		if !rt.TryClaim(result) {
+			return cursor.Err()
+		}
 
-func mergeFilters(idFilter bson.M, customFilter bson.M) bson.M {
-	if len(idFilter) == 0 {
-		return customFilter
+		emit(value)
 	}
 
-	if len(customFilter) == 0 {
-		return idFilter
-	}
+	result := cursorResult{isExhausted: true}
+	rt.TryClaim(result)
```

Comment:
I added this extra `TryClaim` invokation to be able to mark the tracker as done in the event that the cursor is exhausted before the expected count of documents have been claimed. This may happen if documents within the restriction changed, e.g. were deleted or modified to no longer match the filter, since the total number of documents within the restriction was counted upon restriction creation. I preferred to still have the DoFn succeed in this case.

My first idea involved passing only the ID as the pos argument to TryClaim, and returning `sdf.StopProcessing()` after the cursor had been exhausted, inspired by the [example](https://beam.apache.org/documentation/programming-guide/#user-initiated-checkpoint) in the programming guide. However I noticed user initiated checkpointing is less supported among the runners and also my pipeline was interpreted as a streaming job in Dataflow so I was unsure of whether that was the right approach to use and decided to go with this less elegant way instead.

Reply:
I think you're right in this case. 

Formally, the promise around TryClaim is "if TryClaim returns successfully, I will process the claimed position", so if the underlying data/job mapping, changes to invalidate the restriction, it's reasonable to simply discard previously unknown input, since it wasn't originally there to start with. 

As for process continuations, while they *technically* can be used for batch WRT the global window, you're right that support for them is largely for handling streaming, which is how Dataflow interprets transforms that can return process continuations.  It's possible it's a bug in the Go SDKs interpretation however, and it's the combination of a  Process Continuation and an Unbounded Restriction that should lead to a streaming interpretation but that would need to be vetted against existing runners. The documentation around that part of the model for SDK developers is spotty at best unfortunately.

-------

WRT actual splitting behavior, I know Dataflow is currently the only runner that does dynamic splitting. I believe both Flink and Spark will do the initial splitting stages, which amounts to each element will get a split of their initial restriction after creating the restriction.

This is typically fine. Dynamic splitting or Liquid Sharding shows most of it's benefits in avoiding stragglers so that all the work ends closer together, providing better utilization of workers, and faster completion of a job.

Reply:
OK seeing everything in context, I agree 100% this is the correct move because the tracker is managing itself around the external cursor, and can't do the more elegant approach. Ideally, the "last" call to TryClaim that returns true will automatically ensure subsequent claim calls return false.

"This is the last position. I'll return true, but set my state as stopped for future iterations".